src/cluster_asm.c

```diff
@@ -2722,32 +2900,133 @@ void asmTrimJobProcessPending(void) {
+    if (!server.cluster_enabled || !clusterNodeIsMaster(getMyClusterNode())) return;
 
     size_t num_keys = 0;
-    slotRangeArray *sra = NULL;
-    for (int i = 0; i < CLUSTER_SLOTS; i++) {
-        /* Skip slots that are owned, empty, or already in a trim job. */
-        if (clusterIsMySlot(i) ||
-            kvstoreDictSize(server.db[0].keys, i) == 0 ||
-            isSlotInTrimJob(i))
-        {
-            continue;
+    slotRangeArray *trim_sra = NULL;
+    for (int i = 0; i < sra->num_ranges; i++) {
+        for (int j = sra->ranges[i].start; j <= sra->ranges[i].end; j++) {
+            if (clusterIsMySlot(j) ||
+                kvstoreDictSize(server.db[0].keys, j) == 0 ||
+                isSlotInTrimJob(j))
+            {
+                continue;
+            }
+
+            trim_sra = slotRangeArrayAppend(trim_sra, j);
+            num_keys += kvstoreDictSize(server.db[0].keys, j);
         }
-        sra = slotRangeArrayAppend(sra, i);
-        num_keys += kvstoreDictSize(server.db[0].keys, i);
     }
-    if (!sra) return;
+    if (!trim_sra) return;
 
-    sds str = slotRangeArrayToString(sra);
+    sds str = slotRangeArrayToString(trim_sra);
     serverLog(LL_NOTICE,
               "Detected keys in slots that does not belong to this node. "
               "Scheduling trim for %zu keys in slots: %s", num_keys, str);
     sdsfree(str);
 
-    asmTrimJobSchedule(sra);
-    slotRangeArrayFree(sra);
+    asmTrimJobSchedule(trim_sra);
+    slotRangeArrayFree(trim_sra);
+}
+
+/* Handle the orphaned master task when the master of this node is changed or
+ * this node promotes to master. And trim unowned slots when the import task
+ * is failed and trim_slots argument is set. */
+void asmHandleOrphanedMasterTask(int trim_slots) {
+    if (!server.cluster_enabled) return;
+
+    asmTask *task = asmManager->master_task;
+    if (task == NULL) return;
+    serverAssert(task->operation == ASM_IMPORT);
+
+    sds slot_ranges_str = slotRangeArrayToString(task->slot_ranges);
+    serverLog(LL_WARNING, "Failed import task from old master. id=%s, slots=%s",
+                           task->id, slot_ranges_str);
+    sdsfree(slot_ranges_str);
+
+    /* Check if there is an ASM task that master did not finish. */
+    if (task->state != ASM_DONE && task->state != ASM_FAILED) {
+        /* Mark the task as failed and notify the replicas. */
+        task->state = ASM_FAILED;
+        asmNotifyStateChange(task, ASM_EVENT_IMPORT_FAILED);
+    }
+
+    /* Trim the slots if the import task is failed. */
+    if (trim_slots && task->state == ASM_FAILED)
+        asmTrimSlotsIfNotOwned(task->slot_ranges);
+
+    /* Clear the master task since it is not the master anymore. */
+    asmTaskFree(asmManager->master_task);
+    asmManager->master_task = NULL;
+}
+
+/* Handle the orphaned master task when the master of this node is changed. */
+void asmHandleMasterChange(void) {
+    asmHandleOrphanedMasterTask(0);
```

Comment:
why do we finish task when the master change? 

more importantly, let's say we were in the middle of a task. We raise IMPORT_STARTED. Then, the master has changed. We raise IMPORT_FAILED. Then, the new master POV, the task might be going on successfully.  So, at some point, it might send "asm-task" with done state. Then, we'll raise IMPORT_COMPLETED again without IMPORT_STARTED first?